### PR TITLE
Sparse slice with step

### PR DESCRIPTION
**Pull request by ChienliMa**

### CODE REVIEW

**Comment by nouiz** [comment]
travis failed with this error:
```
======================================================================

ERROR: test_GetItem2D (theano.sparse.tests.test_basic.Test_getitem)

----------------------------------------------------------------------

Traceback (most recent call last):

File "/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/sparse/tests/test_basic.py", line 2132, in test_GetItem2D

r1 = f1(vx, m, n, p, q, j, k)

File "/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/compile/function_module.py", line 606, in __call__

storage_map=self.fn.storage_map)

File "/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/compile/function_module.py", line 595, in __call__

outputs = self.fn()

File "/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/gof/op.py", line 748, in rval

r = p(n, [x[0] for x in i], o)

File "/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/sparse/basic.py", line 1280, in perform

out[0] = x[start1:stop1:step1, start2:stop2:step2]

File "/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/scipy/sparse/csc.py", line 147, in __getitem__

return self.T[col,row].T

File "/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/scipy/sparse/csr.py", line 241, in __getitem__

return self._get_submatrix(row, col) #[1:2,j]

File "/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/scipy/sparse/csr.py", line 380, in _get_submatrix

i0, i1 = process_slice( row_slice, M )

File "/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/scipy/sparse/csr.py", line 351, in process_slice

raise ValueError('slicing with step != 1 not supported')

ValueError: slicing with step != 1 not supported

Apply node that caused the error: GetItem2d(SparseVariable{csc,float64}, a, b, e, c, d, f)

Inputs types: [Sparse[float64, csc], TensorType(int32, scalar), TensorType(int32, scalar), TensorType(int32, scalar), TensorType(int32, scalar), TensorType(int32, scalar), TensorType(int32, scalar)]

Inputs shapes: [(100, 97), (), (), (), (), (), ()]

Inputs strides: ['No strides', (), (), (), (), (), ()]

Inputs values: ['not shown', array(1, dtype=int32), array(5, dtype=int32), array(2, dtype=int32), array(10, dtype=int32), array(15, dtype=int32), array(3, dtype=int32)]

Backtrace when the node is created:

File "/home/travis/miniconda/envs/pyenv/lib/python2.6/unittest.py", line 300, in __call__

return self.run(*args, **kwds)

File "/home/travis/miniconda/envs/pyenv/lib/python2.6/unittest.py", line 279, in run

testMethod()

File "/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/sparse/tests/test_basic.py", line 2131, in test_GetItem2D

f1 = theano.function([x, a, b, c, d, e, f], x[a:b:e, c:d:f])

File "/home/travis/miniconda/envs/pyenv/lib/python2.6/site-packages/theano/sparse/basic.py", line 344, in __getitem__

ret = get_item_2d(self, args)
```

Is this because we use an old version of scipy on travis? If so, we should skip the test with scipy version is too old. Or make the test catch that specific error.

**Comment by nouiz** [comment]
I just got a few small comments. Thanks.

**Comment by ChienliMa** [comment]
This test case was passed in my computer with Scipy 0.14.1.
Without knowing when did scipy start supporting slicing with step in sparse matrix, I think I'd better catch that error.

File: theano/sparse/basic.py [hunk 1]
```diff
@@ -1219,15 +1219,22 @@ def make_node(self, x, index):
...
-                # Else, they should be converted to Tensor Variables of
-                # dimension 1 and int/uint dtype.
+                step = ind.step
+                # If start or stop or step are None, make them a Generic 
+                # constant. Else, they should be converted to Tensor Variables
+                # of dimension 1 and int/uint dtype.
+                if ind.step is  None or ind.step == 1:
+                    step = generic_None
+                else:
+                    if not isinstance(step, gof.Variable):
```
Comment (nouiz) [inline theano/sparse/basic.py]: Travis use scipy 0.11. This version raise an error is step isn't 1. We know that older version was just ignoring the step when it wasn't None. Just to help the user, can you assert here that scipy version is >= 0.11 ? That way we will be sure one error will be raised to the user (Theano one or scipy one) without asking us to test all scipy version.

**Comment by ChienliMa** [comment]
I had a test with Scipy 0.11.0, the value errors were not raised when testing with:
```python
 self.assertRaises(ValueError, x.__getitem__, (slice(a, b, 4), slice(c, d)))
```
but this error was raise if  I set step in previous test not equal 1.

**Comment by ChienliMa** [comment]
OIC, I should raise this value myself.

File: theano/sparse/tests/test_basic.py [hunk 2]
```diff
@@ -2111,23 +2115,36 @@ def test_GetItem2D(self):
...
-            assert numpy.all(t1.toarray() == r1.toarray())
+            if is_supported_version:
+                f1 = theano.function([x, a, b, c, d, e, f], x[a:b:e, c:d:f])
+                r1 = f1(vx, m, n, p, q, j, k)
+            else:
+                f1 = theano.function([x, a, b, c, d], x[a:b, c:d])
+                r1 = f1(vx, m, n, p, q)
+                t1 = vx[m:n, p:q]
+                assert r1.shape == t1.shape
+                assert numpy.all(t1.toarray() == r1.toarray())
```
Comment (nouiz) [inline theano/sparse/tests/test_basic.py]: The assert have bad indentation.

File: theano/sparse/tests/test_basic.py [hunk 3]
```diff
@@ -2160,32 +2177,43 @@ def test_GetItem2D(self):
...
+                r4 = f4(vx, m, n, j)
+                t4 = vx[m:n:j]
+                assert r4.shape == t4.shape
+                assert numpy.all(t4.toarray() == r4.toarray())
+            else:
+                f4 = theano.function([x, a, b], x[a:b])
+                r4 = f4(vx, m, n)
+                t4 = vx[m:n]
+                assert r4.shape == t4.shape
+                assert numpy.all(t4.toarray() == r4.toarray())
```
Comment (nouiz) [inline theano/sparse/tests/test_basic.py]: Remove duplicate assert, just keep one after the ifelse at the same level as the if.

**Comment by ChienliMa** [comment]
The travis build fell. May I push another commit to restart it, or there is another way?

**Comment by nouiz** [comment]
I restarted it. You can close and reopen the PR to do the same.

**Comment by nouiz** [comment]
thanks.

**Pull request merged**